[wallet] Add wallet name to log messages

PierreRochard opened the pull request:
After multiple wallets became supported, wallet-related log messages became ambiguous as to which wallet they were being emitted by.

This pull request adds a `CWallet::WalletLogPrintf` function to be used when logging wallet-specific events. This function prepends the wallet's name to the log message and forwards it to `LogPrintf`

fixes #11317

PierreRochard commented:
Thank you @dooglus for bringing attention to this in #11317 and opening #11320. 

This is #11320 with the following changes:
 * rebased on master
 * added the wallet name in three additional logging messages for rpcdump.cpp's importwallet function

In #11320 @promag and @luke-jr suggested approaches to reducing the repetition of `"[%s]" GetName()` boilerplate. @jimpo, in the context of #12954, what are your thoughts?

This is outside of the scope of this PR and may be controversial, but I do think we would benefit from automated testing of log messages. This could either be done by checking the `debug.log` of nodes in existing tests, or as a separate test script.

Open questions:

1. There are logging calls in two `WalletBatch` methods where it's not immediately apparent to me how to get the wallet's name: `WalletBatch::FindWalletTx` and `WalletBatch::RecoverKeysOnlyFilter`. Is there something I'm not seeing? If the only solution is refactoring then I'll leave them alone in this PR.

2. `CWallet`'s `m_name` defaults to an empty string, so we end up with lots of empty square brackets in the log messages. Should we have a default wallet name?

jimpo commented:
I luke-jr's suggestion of `CWallet::LogPrintf`. The main benefit is not even reducing boilerplate, it's to reduce the amount of refactoring required if someone decides that there is a better way of presenting the wallet info on the log line. Also, I think it'll be more obvious that way when people add new log statements in wallet files as compared to noticing the extra log argument and having to realize that it's a standard.

With regards to #12954, I don't see any issues. Won't even be a git conflict, I don't think.

This is an example of a case where [structured logging](https://stackify.com/what-is-structured-logging-and-why-developers-need-it/) might be useful (being able to tag log statments with key-value pairs in a not-ad-hoc way). I know the Bitcoin ABC team is thinking about this.

MarcoFalke commented:
Agree that it makes sense to move this to `CWallet::LogPrintf` instead.

MarcoFalke commented on src/wallet/rpcdump.cpp:
```diff
@@ -554,7 +554,7 @@ UniValue importwallet(const JSONRPCRequest& request)
         // Use uiInterface.ShowProgress instead of pwallet.ShowProgress because pwallet.ShowProgress has a cancel button tied to AbortRescan which
         // we don't want for this progress bar shoing the import progress. uiInterface.ShowProgress does not have a cancel button.
-        uiInterface.ShowProgress(_("Importing..."), 0, false); // show progress dialog in GUI
+        uiInterface.ShowProgress(strprintf(_("[%s] Importing..."), pwallet->GetName()), 0, false); // show progress dialog in GUI
```
Weird ascii characters should not be translated. You can keep the `_("Importing...")` and do the string manipulation afterward

jnewbery commented:
Concept ACK. I agree that `CWallet::LogPrintf` would be useful.

PierreRochard committed:
[wallet] Add wallet name to log messages

After multiple wallets became supported, wallet-related log messages
became ambiguous as to which wallet they were being emitted by.

fixes #11317

PierreRochard commented:
Unfortunately `CWallet::LogPrintf` doesn't play nice with the `LogPrintf` macro so I created a `CWallet::WalletLogPrintf` function, using a variadic template to forward args to the macro. Creative solutions to avoid the double namespacing are welcome

https://github.com/PierreRochard/bitcoin/commit/45343a251306000501654d1dc015dd88501e271d → https://github.com/PierreRochard/bitcoin/commit/77152113c9939d4ba7b6e8f22ec4d07703d1891b force push diff [here](https://pierrerochard.github.io/bitcoin/diffs/pr_log-wallet-name_1__pr_log-wallet-name_2.html#d2h-048665)

jnewbery commented on src/wallet/wallet.h:
```diff
@@ -1198,6 +1198,11 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
+
+    template<typename... Params>
+    void WalletLogPrintf(std::string fmt, Params... parameters) const {
+        LogPrintf(("[" + GetName() + "] " + fmt).c_str(), parameters...);
```
This results in empty brackets at the start of the log if the default is being used, eg:

```
2018-06-21T14:12:32.172374Z init message: [] Loading wallet...
2018-06-21T14:12:32.174733Z [] nFileVersion = 169900
2018-06-21T14:12:32.174776Z [] Keys: 0 plaintext, 0 encrypted, 0 w/ metadata, 0 total. Unknown wallet records: 0
2018-06-21T14:12:32.175201Z [] Performing wallet upgrade to 169900
```

Any thoughts about only adding the prefix if the wallet name is non-empty?

PierreRochard replied:
Agree empty is ugly, I'm ended up changing it to be like the GUI and put `default wallet` there instead, so that it's explicit.

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -4057,7 +4056,7 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(const std::string& name,
     }
 
-    uiInterface.InitMessage(_("Loading wallet..."));
+    uiInterface.InitMessage(strprintf(_("[%s] Loading wallet..."), walletFile));
```
nit: This looks a bit weird in the logs compared to the other messages:

```
2018-06-21T14:12:32.894599Z init message: [w1] Loading wallet...
```

(ie the `[<wallet_name>]` prefix is not at the start of the log). Maybe just change to:

```
    uiInterface.InitMessage(strprintf(_("Loading wallet %s ..."), walletFile));
```

PierreRochard replied:
Fixed

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -4098,12 +4097,12 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(const std::string& name,
+            walletInstance->WalletLogPrintf("Allowing wallet upgrade up to %i\n", nMaxVersion);
         if (nMaxVersion < walletInstance->GetVersion())
         {
             InitError(_("Cannot downgrade wallet"));
```
There are a couple of logs lower down that you've missed:

- `            LogPrintf("Upgrading wallet to HD\n");` (L4131)
- `            LogPrintf("Upgrading wallet to use HD chain split\n");` (L4143)

PierreRochard replied:
Fixed

MarcoFalke replied:
Is this one fixed?

jnewbery commented on src/wallet/feebumper.cpp:
```diff
@@ -195,7 +195,7 @@ Result CreateTransaction(const CWallet* wallet, const uint256& txid, const CCoin
     poutput->nValue -= nDelta;
     if (poutput->nValue <= GetDustThreshold(*poutput, GetDiscardRate(*wallet, ::feeEstimator))) {
-        LogPrint(BCLog::RPC, "Bumping fee and discarding dust output\n");
+        LogPrint(BCLog::RPC, "[%s] Bumping fee and discarding dust output\n", wallet->GetName());
```
I think this should just be changed to a `WalletLogPrintf`. I don't think this needs to be in the RPC category.

PierreRochard replied:
Fixed

jnewbery reviewed (commented):
Tested ACK 77152113c9939d4ba7b6e8f22ec4d07703d1891b.

A few nits inline.

The log format conflicts slightly with #13168, which prefixes the log with `[<thread_name>]`. It'll make the logic a little more tricky for log parsing tools to deal with the possibility of one or two bracketed prefixes, but it's probably not that big a deal.

I agree with @jimpo that structured logging could be a good way to go in future. That'd obviously be a separate PR.

PierreRochard commented:
Thanks for the review John, I've addressed your nits

https://github.com/PierreRochard/bitcoin/commit/77152113c9939d4ba7b6e8f22ec4d07703d1891b → https://github.com/PierreRochard/bitcoin/commit/553b1a7f815ec39dc30c65f22aa2f9f6cae41c25 force push diff [here](https://pierrerochard.github.io/bitcoin/diffs/pr_log-wallet-name_2__pr_log-wallet-name_3.html)

l2a5b1 reviewed (commented):
Nice job @PierreRochard! Tested ACK 553b1a7. Feedback is not critical, feel free to ignore it.

Empact commented on src/wallet/wallet.h:
```diff
@@ -1198,6 +1198,11 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
     /** Whether a given output is spendable by this wallet */
     bool OutputEligibleForSpending(const COutput& output, const CoinEligibilityFilter& eligibility_filter) const;
+
+    template<typename... Params>
```
I also prefer this factoring. Maybe a method like `CWallet::DisplayName` is called for - the concept is used in e.g. https://github.com/bitcoin/bitcoin/blob/88a15ebc8d317a6fd4851adb344ff944d497284c/src/qt/bitcoingui.cpp#L509

PierreRochard replied:
@251Labs I was hoping that there'd be an approach so that the logging user can call `LogPrintf` and not have to think about the context - the macro or some other magic would automatically add the wallet name if it's available in that context. Ultimately I think that kind of improvement would be part of a wider (structured) logging refactoring.

@Empact agree, I've factored out a `CWallet::GetDisplayName` and added calls to it in relevant parts

251Labs replied:
Thanks @PierreRochard, yeah you're absolutely right about that. 

I was thinking about something more straightforward. 

The gist of my suggestion was to *just* extract (or remove) `WalletLogPrintf` from the wallet for the advantages that this may provide.

When extracted the signature of `WalletLogPrintf` could look something like this:

```c++
template<typename... Params> 
void WalletLogPrintf(const CWallet* wallet, const std::string& fmt, const Params&... parameters) {
    ...
};
```
and would be used like:
```c++
WalletLogPrintf(pWallet, "#reckless");
```

Imo `WalletLogPrintf` meets the recommendation: “A function that can be simply, elegantly, and efficiently implemented as a freestanding function (that is, as a nonmember function) should be implemented outside the class. ” from Stroustrup.

Anyway, this comment is just to clarify what I meant. It is not critical and it's probably best to ignore it :)

Empact commented on src/wallet/rpcdump.cpp:
```diff
@@ -560,7 +560,7 @@ UniValue importwallet(const JSONRPCRequest& request)
         // Use uiInterface.ShowProgress instead of pwallet.ShowProgress because pwallet.ShowProgress has a cancel button tied to AbortRescan which
         // we don't want for this progress bar showing the import progress. uiInterface.ShowProgress does not have a cancel button.
-        uiInterface.ShowProgress(_("Importing..."), 0, false); // show progress dialog in GUI
+        uiInterface.ShowProgress(strprintf(_("[%s] Importing..."), pwallet->GetName()), 0, false); // show progress dialog in GUI
```
Seems like a need for `GetDisplayName` here.

PierreRochard replied:
Fixed

achow101 commented:
Needs rebase

sipa commented:
Needs trivial rebase after #13500.

251Labs commented on src/wallet/wallet.cpp:
```diff
@@ -1669,8 +1668,8 @@ void CWalletTx::GetAmounts(std::list<COutputEntry>& listReceived,
-            LogPrintf("CWalletTx::GetAmounts: Unknown transaction type found, txid %s\n",
-                     this->GetHash().ToString());
+            pwallet->WalletLogPrintf("CWalletTx::GetAmounts: Unknown transaction type found, txid %s\n",
+                                     this->GetHash().ToString());
```
nit: whitespace (alignment is off by one space).

PierreRochard replied:
Fixed

achow101 commented on src/wallet/wallet.h:
```diff
@@ -1185,6 +1185,15 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
 
     /** Whether a given output is spendable by this wallet */
     bool OutputEligibleForSpending(const COutput& output, const CoinEligibilityFilter& eligibility_filter) const;
+    
```
Extra whitespace on this line is causing the linter to fail.

PierreRochard replied:
Ah thank you, I'll take it as an opportunity to add comments, that should fix it

achow101 commented:
utACK fb71743dbbd188ec3796f6e489f0a73e328bb7cb

PierreRochard commented:
Thank you for the reviews and comments @empact, @251Labs, and @achow101. 

https://github.com/PierreRochard/bitcoin/commit/553b1a7f815ec39dc30c65f22aa2f9f6cae41c25 → https://github.com/PierreRochard/bitcoin/commit/6d77ba31e760c0f1806b721e34437724da199955 force push diff [here](https://pierrerochard.github.io/bitcoin/diffs/pr_log-wallet-name_3__pr_log-wallet-name_4.html)

PierreRochard commented:
Trivial rebase on to https://github.com/bitcoin/bitcoin/pull/9662

skeees commented:
utACK dd67b82
on an unrelated note the force push diffs are great (i'm guessing you have some automated solution that does those) - would be very cool functionality to put into a bot of some sort

MarcoFalke commented:
@skeees I presume the diff would only be useful if you reviewed the pre-force-push diff as well. In which case you could just do `git diff $pre_force_push_commit $new_commit` locally (after fetching the objects)

PierreRochard commented:
Thank you for the review @skeees, I've been manually experimenting with different approaches and I think I've found a good way to do it with interdiff + diff2html. I'll be automating it and integrating it with bitcoinacks.com, if people find it useful I could work with @MarcoFalke to make a bot.

@MarcoFalke agree it may only be a marginal improvement

skeees commented:
yah - its the sort of thing where the people i know who are pro reviewers have built their own tools to do this already. ideally github would have this, but in lieu of that would be an excellent addition to bitcoinacks when you've got the time :)

jnewbery commented on src/wallet/wallet.h:
```diff
@@ -1206,6 +1206,18 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
+    /** Prepends the wallet name in logging output to ease debugging in multi-wallet use cases */
+    template<typename... Params>
+    void WalletLogPrintf(std::string fmt, Params... parameters) const {
+        LogPrintf((GetDisplayName() + " " + fmt).c_str(), parameters...);
```
sanitize inputs please :)

```
/home/ubuntu/bitcoin
→ bcli createwallet "a%ib"
{
  "name": "a%ib",
  "warning": ""
}
tail -30 ../.bitcoin/regtest/debug.log
2018-07-27T14:36:38Z tor: Not connected to Tor control port 127.0.0.1:9051, trying to reconnect
2018-07-27T14:36:40Z tor: Error connecting to Tor control socket
2018-07-27T14:36:40Z tor: Not connected to Tor control port 127.0.0.1:9051, trying to reconnect
2018-07-27T14:36:44Z tor: Error connecting to Tor control socket
2018-07-27T14:36:44Z tor: Not connected to Tor control port 127.0.0.1:9051, trying to reconnect
2018-07-27T14:36:45Z Received a POST request for / from 127.0.0.1:59554
2018-07-27T14:36:45Z ThreadRPCServer method=createwallet user=bitcoinrpc
2018-07-27T14:36:45Z Using BerkeleyDB version Berkeley DB 4.8.30: (April  9, 2010)
2018-07-27T14:36:45Z Using wallet wallet.dat
2018-07-27T14:36:45Z BerkeleyEnvironment::Open: LogDir=/home/ubuntu/.bitcoin/regtest/wallets/a%ib/database ErrorFile=/home/ubuntu/.bitcoin/regtest/wallets/a%ib/db.log
2018-07-27T14:36:45Z init message: Loading wallet a%ib...
2018-07-27T14:36:45Z Error "tinyformat: Too many conversion specifiers in format string" while formatting log message: [a%ib] nFileVersion = %d
2018-07-27T14:36:45Z Error "tinyformat: Too many conversion specifiers in format string" while formatting log message: [a%ib] Keys: %u plaintext, %u encrypted, %u w/ metadata, %u total. Unknown wallet records: %u
2018-07-27T14:36:45Z Error "tinyformat: Too many conversion specifiers in format string" while formatting log message: [a%ib] Performing wallet upgrade to %i
2018-07-27T14:36:46Z Error "tinyformat: Too many conversion specifiers in format string" while formatting log message: [a%ib] keypool added %d keys (%d internal), size=%u (%u internal)
2018-07-27T14:36:46Z Error "tinyformat: Too many conversion specifiers in format string" while formatting log message: [a%ib] wallet      %15dms
2018-07-27T14:36:46Z Error "tinyformat: Too many conversion specifiers in format string" while formatting log message: [a%ib] setKeyPool.size() = %u
2018-07-27T14:36:46Z Error "tinyformat: Too many conversion specifiers in format string" while formatting log message: [a%ib] mapWallet.size() = %u
2018-07-27T14:36:46Z Error "tinyformat: Too many conversion specifiers in format string" while formatting log message: [a%ib] mapAddressBook.size() = %u
2018-07-27T14:36:49Z Flushing wallet.dat
2018-07-27T14:36:49Z Flushed wallet.dat 4ms
2018-07-27T14:36:49Z tor: Error connecting to Tor control socket
2018-07-27T14:36:49Z tor: Not connected to Tor control port 127.0.0.1:9051, trying to reconnect
2018-07-27T14:36:56Z tor: Error connecting to Tor control socket
2018-07-27T14:36:56Z tor: Not connected to Tor control port 127.0.0.1:9051, trying to reconnect
2018-07-27T14:37:08Z tor: Error connecting to Tor control socket
2018-07-27T14:37:08Z tor: Not connected to Tor control port 127.0.0.1:9051, trying to reconnect
2018-07-27T14:37:25Z tor: Error connecting to Tor control socket
2018-07-27T14:37:25Z tor: Not connected to Tor control port 127.0.0.1:9051, trying to reconnect
2018-07-27T14:37:36Z Adding fixed seed nodes as DNS doesn't seem to be available.
```

MarcoFalke replied:
This would probably conflict with unicode support in wallet file names, no?

MarcoFalke replied:
Should probably escape or quote instead.

PierreRochard replied:
@MarcoFalke I definitely agree that escape or quote would be better than sanitize (which simply drops characters), however the only existing escape functions I found were in leveldb, qt, and univalue. Any advice before I try my hand at adding one?

skeees replied:
Rather than adding an escape fn another option is to just 

>LogPrintf("%s " + fmt, GetDisplayName(), parameters...)

MarcoFalke replied:
Good point!

PierreRochard commented:
Thank you for catching that @jnewbery!

https://github.com/PierreRochard/bitcoin/commit/dd67b82bbeba948823224d9ab6c2aa2d24730e21 → https://github.com/PierreRochard/bitcoin/commit/237a0f87c9841c46276b04b1f947db9fcd10221dforce push diff [here](https://pierrerochard.github.io/bitcoin/diffs/pr_log-wallet-name_5__pr_log-wallet-name_6.html)

MarcoFalke commented:
There might be some more instances after you rebase on master:

```sh
git grep LogPrint ./src/wallet ':(exclude)src/wallet/db.cpp' ':(exclude)src/wallet/init.cpp' ':(exclude)src/wallet/walletdb.cpp' ':(exclude)src/wallet/crypter.cpp' |grep -v WalletLogPrintf| grep -v GetDisplayName

MarcoFalke commented on src/wallet/rpcdump.cpp:
```diff
@@ -560,7 +560,7 @@ UniValue importwallet(const JSONRPCRequest& request)
         // Use uiInterface.ShowProgress instead of pwallet.ShowProgress because pwallet.ShowProgress has a cancel button tied to AbortRescan which
         // we don't want for this progress bar showing the import progress. uiInterface.ShowProgress does not have a cancel button.
-        uiInterface.ShowProgress(_("Importing..."), 0, false); // show progress dialog in GUI
+        uiInterface.ShowProgress(strprintf(_("%s Importing..."), pwallet->GetDisplayName()), 0, false); // show progress dialog in GUI
```
If you prefix this, it would be better to not change the translation string. I.e. `strprintf("%s " + _("Imp...`

laanwj replied:
agree

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -1728,11 +1727,11 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, CBlock
     {
         fAbortRescan = false;
-        ShowProgress(_("Rescanning..."), 0); // show rescan progress in GUI as dialog or on splashscreen, if -rescan on startup
+        ShowProgress(strprintf(_("%s Rescanning..."), GetDisplayName()), 0); // show rescan progress in GUI as dialog or on splashscreen, if -rescan on startup
```
Same here and all cases below ...

laanwj commented:
lightly-tested ACK 909f54c80abb7195c2e82c6e06c414f4526a339e

jnewbery reviewed (commented):
post-merge tested ACK 909f54c80abb7195c2e82c6e06c414f4526a339e. A few nits inline.

PierreRochard commented on src/wallet/wallet.cpp:
```diff
@@ -4006,7 +4005,7 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(const std::string& name,
     }
 
-    uiInterface.InitMessage(_("Loading wallet..."));
+    uiInterface.InitMessage(strprintf(_("Loading wallet %s..."), walletFile));
```
Yes, reverted in https://github.com/bitcoin/bitcoin/pull/13911 and I'll open up a post-0.17 PR for `InitMessage`, `InitWarning`, and `InitError` messages

PierreRochard commented on src/wallet/wallet.cpp:
```diff
@@ -4218,7 +4217,7 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(const std::string& name,
     walletInstance->m_signal_rbf = gArgs.GetBoolArg("-walletrbf", DEFAULT_WALLET_RBF);
 
-    LogPrintf(" wallet      %15dms\n", GetTimeMillis() - nStart);
+    walletInstance->WalletLogPrintf("wallet      %15dms\n", GetTimeMillis() - nStart);
```
Agree, fixed in https://github.com/bitcoin/bitcoin/pull/13911

PierreRochard commented on src/wallet/wallet.cpp:
```diff
@@ -4271,7 +4270,7 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(const std::string& name,
             walletInstance->ScanForWalletTransactions(pindexRescan, nullptr, reserver, true);
         }
-        LogPrintf(" rescan      %15dms\n", GetTimeMillis() - nStart);
+        walletInstance->WalletLogPrintf("rescan      %15dms\n", GetTimeMillis() - nStart);
```
Agree, fixed in https://github.com/bitcoin/bitcoin/pull/13911
